File: pd/31-03-2020/knight.pl
:- module(knight, [
    jump/1,
    on_board/1,
    move_knight/2,
    reach_in_3/2
]).

jump((X,Y)) :-
    member(X, [-2,-1,1,2]),
    member(Y, [-2,-1,1,2]),
    3 =:= abs(X) + abs(Y). 

on_board((X,Y)) :-
    X >= 1, X =< 8,
    Y >= 1, Y =< 8.

move_knight((X0,Y0), (X1,Y1)) :-
    jump((A,B)),
    X1 is X0+A,
    Y1 is Y0+B,
    on_board((X1,Y1)).

reach_in_3(Pos0, Pos3) :-
    move_knight(Pos0, Pos1),
    move_knight(Pos1, Pos2),
    move_knight(Pos2, Pos3).

File: pd/31-03-2020/Knight.hs
module Knight(
    KnightPos,
    jump,
    onBoard,
    moveKnight,
    in3,
    reachIn3
) where

import Control.Monad(guard)
type KnightPos = (Int,Int)

jump :: [KnightPos]
jump = do x <- [-2,-1,1,2]
          y <- [-2,-1,1,2]
          guard (3 == abs x + abs y)
          return (x,y)

onBoard :: KnightPos -> Bool
onBoard (x,y) = x `elem` [1..8] && y `elem` [1..8]

moveKnight :: KnightPos -> [KnightPos]
moveKnight (x,y) = do (a,b) <- jump
                      guard $ onBoard (a+x,b+y)
                      return (a+x,b+y)

in3 :: KnightPos -> [KnightPos]
in3 pos0 = do pos1 <- moveKnight pos0
              pos2 <- moveKnight pos1
              moveKnight pos2

reachIn3 :: KnightPos -> KnightPos -> Bool
reachIn3 from to = to `elem` in3 from